src/machines/activity_state_machine.rs

```diff
@@ -417,9 +537,73 @@ pub(super) struct TimedOut {}
 #[derive(Default, Clone)]
 pub(super) struct Canceled {}
 
+fn create_request_cancel_activity_task_command(
+    dat: SharedState,
+    next_state: ActivityMachineState,
+) -> TransitionResult<ActivityMachine> {
+    // createRequestCancelActivityTaskCommand
+    let cmd = Command {
+        command_type: CommandType::RequestCancelActivityTask as i32,
+        attributes: Some(
+            RequestCancelActivity {
+                scheduled_event_id: dat.scheduled_event_id,
+                activity_id: dat.attrs.activity_id,
+            }
+            .into(),
+        ),
+    };
+    ActivityMachineTransition::ok(
+        vec![ActivityMachineCommand::RequestCancellation(cmd)],
+        next_state,
+    )
+}
+
+/// Notifies lang side that activity has been cancelled by sending a failure with cancelled failure as a cause.
+/// Optional cancelled_event, if passed, is used to supply event IDs.
+fn notify_canceled(
```

Comment:
I think we should give this a better name now that we understand it better. Maybe `tell_lang_activity_cancelled`

The docstring should also mention that we transition to the provided state, and the "tell lang" part is the `Fail` command that we put in the transition.

Reply:
Done